src/messages/proposals.rs

```diff
@@ -293,38 +363,49 @@
         }
         true
     }
+    /// Returns proposal for a given proposal ID
+    pub(crate) fn get(&self, proposal_id: &ProposalID) -> Option<&QueuedProposal> {
+        self.queued_proposals.get(proposal_id)
+    }
     /// Add a new `QueuedProposal` to the queue
-    pub(crate) fn add(&mut self, queued_proposal: QueuedProposal) {
+    pub(crate) fn add(&mut self, queued_proposal: QueuedProposal<'a>) {
         self.queued_proposals
             .entry(queued_proposal.proposal_id.clone())
             .or_insert(queued_proposal);
     }
     /// Retains only the elements specified by the predicate
     pub(crate) fn retain<F>(&mut self, f: F)
     where
-        F: FnMut(&ProposalID, &mut QueuedProposal) -> bool,
+        F: FnMut(&ProposalID, &mut QueuedProposal<'a>) -> bool,
     {
         self.queued_proposals.retain(f);
     }
     /// Gets the list of all `ProposalID`
-    pub(crate) fn proposal_id_list(&self) -> Vec<ProposalID> {
-        self.queued_proposals.keys().into_iter().cloned().collect()
+    pub(crate) fn commit_list(&self) -> Vec<ProposalOrRef> {
+        self.queued_proposals
+            .iter()
+            .map(
+                |(proposal_id, queued_proposal)| match queued_proposal.por_type {
+                    ProposalOrRefType::Proposal => {
+                        ProposalOrRef::Proposal(queued_proposal.proposal.clone())
+                    }
+                    ProposalOrRefType::Reference => ProposalOrRef::Reference(proposal_id.clone()),
+                    _ => {
+                        panic!("Library error. Wrong Queued Proposals type.")
```

Comment:
I don't see that we check the type when inserting proposal into the queue. So there's a real chance that we run into this panic. Either make sure that we check the type when inserting or return an error here instead.